crates/matrix-sdk/tests/integration/encryption.rs

```diff
@@ -0,0 +1,237 @@
+    own_device.verify().await.unwrap();
+    assert!(own_device.is_verified());
+
+    // Bootstrapping signed the user identity.
+    let user_identity = alice.encryption().get_user_identity(&user_id).await.unwrap().unwrap();
+
+    assert_eq!(user_identity.user_id(), user_id);
+    assert!(user_identity.is_verified());
+
+    let master_pub_key = user_identity.master_key();
+    assert_eq!(master_pub_key.user_id(), user_id);
+    assert!(!master_pub_key.keys().is_empty());
+    assert_eq!(master_pub_key.keys().iter().count(), 1);
+
+    // Manually re-verifying doesn't change the outcome.
+    user_identity.verify().await.unwrap();
+    assert!(user_identity.is_verified());
+}
+
+#[async_test]
+async fn test_mutual_verification() {
+    let mut server = MockedServer::new().await;
+
+    let alice_user_id = owned_user_id!("@alice:example.org");
+    let alice_device_id = owned_device_id!("4L1C3");
+    let alice = Client::builder()
+        .homeserver_url(server.server.uri())
+        .server_versions([MatrixVersion::V1_0])
+        .request_config(RequestConfig::new().disable_retry())
+        .build()
+        .await
+        .unwrap();
+    alice
+        .restore_session(MatrixSession {
+            meta: SessionMeta {
+                user_id: alice_user_id.clone(),
+                device_id: alice_device_id.clone(),
+            },
+            tokens: MatrixSessionTokens { access_token: "1234".to_owned(), refresh_token: None },
+        })
+        .await
+        .unwrap();
+
+    let bob = Client::builder()
+        .homeserver_url(server.server.uri())
+        .server_versions([MatrixVersion::V1_0])
+        .request_config(RequestConfig::new().disable_retry())
+        .build()
+        .await
+        .unwrap();
+
+    let bob_user_id = owned_user_id!("@bob:example.org");
+    let bob_device_id = owned_device_id!("B0B0B0B0B");
+    bob.restore_session(MatrixSession {
+        meta: SessionMeta { user_id: bob_user_id.clone(), device_id: bob_device_id.clone() },
+        tokens: MatrixSessionTokens { access_token: "5678".to_owned(), refresh_token: None },
+    })
+    .await
+    .unwrap();
+
+    server.add_known_device(&alice_device_id);
+    server.add_known_device(&bob_device_id);
+
+    let alice_verifies_bob =
+        alice.encryption().get_verification(bob.user_id().unwrap(), "flow_id").await;
+    assert!(alice_verifies_bob.is_none());
+
+    let alice_verifies_bob_request =
+        alice.encryption().get_verification_request(&bob_user_id, "flow_id").await;
+    assert!(alice_verifies_bob_request.is_none());
+
+    let bob_device = alice.encryption().get_device(&bob_user_id, &bob_device_id).await.unwrap();
+    assert!(bob_device.is_none());
+
+    // Have both Alice and Bob bootstrap cross-signing.
+    bootstrap_cross_signing_and_verify_device_test_helper(&alice).await;
+    bootstrap_cross_signing_and_verify_device_test_helper(&bob).await;
+
+    // TODO: continue here with mutual verification using sas
```

Comment:
Yeah that one will be tricky. Perhaps you as a first step you can manually verify the other persons user identity and check that the device is verified as well. That one only requires the signatures to be put into a `/keys/query` after correctly calling `UserIdentity::verify()`.

Reply:
Yeah, I've removed that second test since it's quite incomplete; will add in a followup.